### PR TITLE
Words

### PR DESCRIPTION
**Pull request by shechframgia**
This is for #55689(Word Features)

### CODE REVIEW

File: app/Http/Controllers/WordController.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        return redirect('words');
+    }
+
+    public function destroy(Word $word)
+    {
+        $word->delete();
+        return redirect('words');
+    }
+
+    public function search()
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/WordController.php]: Why did you have empty function?

File: app/Http/routes.php [hunk 2]
```diff
@@ -28,13 +28,16 @@
 
 Route::group(['middleware' => 'web'], function () {
     Route::auth();
+    Route::resourceparameters([
+        'categories' => 'category',
+        'words' => 'word',
+    ]);
     Route::group(['middleware' => 'auth'], function () {
         Route::get('/home', 'HomeController@index');
+        Route::post('/search', 'WordController@search');
```
Comment (nguyenthanhtung88) [inline app/Http/routes.php]: You did not have this feature implement right ?

File: app/Repositories/Eloquents/WordRepository.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        if (!empty($request->file('sound_file'))) {
+            $soundData = $request->file('sound_file');
+            $path = public_path('sounds');
+            $soundName = time() . "." . $soundData->getClientOriginalExtension();
+            $upload = $soundData->move($path, $soundName);
+        } else {
+            $soundName = null;
+        }
+        $data = [
```
Comment (nguyenthanhtung88) [inline app/Repositories/Eloquents/WordRepository.php]: `data` variable name is so general and doesn't make sense, please choose another name like `soundData`

File: database/migrations/2016_03_03_053331_create_categories_table.php [hunk 4]
```diff
@@ -15,6 +15,7 @@ public function up()
         Schema::create('categories', function (Blueprint $table) {
             $table->increments('id');
             $table->string('name');
+            $table->string('image');
```
Comment (nguyenthanhtung88) [inline database/migrations/2016_03_03_053331_create_categories_table.php]: Don't edit old migration file, please make a new migration file for this change

File: resources/views/word/create.blade.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                    {{ Form::text('vietnamese_word', null, ['class' => 'form-control']) }}
+                </div>
+
+                <div class="form-group">
+                    {{ Form::label('japanese_word', trans('text.japanese_word')) }}
+                    {{ Form::text('japanese_word', null, ['class' => 'form-control']) }}
+                </div>
+
+                <div class="form-group">
+                    {{ Form::label('words_option', trans('text.words_option')) }}
```
Comment (nguyenthanhtung88) [inline resources/views/word/create.blade.php]: I think using `word_options` is better

File: resources/views/word/create.blade.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                    {{ Form::text('first_option', null, ['class' => 'form-control', 'placeholder' => trans('text.first_option')]) }} </br>
```
Comment (nguyenthanhtung88) [inline resources/views/word/create.blade.php]: For `word_options` input, with this code for scaling i think is not good.
Image that in the future you will have more options, you have to copy/paste and modify right?
My suggest solution for you:

1. Define a constant for number of word options, assign value is 3 for example
2. In view, make a loop function and render `{{ Form::text('word_options[]', null, ['class' => 'form-control', 'placeholder' => trans('text.word_option')]) }}` => you can try it
3. In controller, you can get `word_options` array, you don't have to take care about how many options of words

File: resources/views/word/wordlist.blade.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+                            </thead>
+                            <tbody>
+                                @if($words->count())
+                                    @foreach($words as $word)
+                                        <tr>
+                                            <td>{{ $word->vietnamese_words }}</td>
+                                            <td>{{ $word->japanese_words }}</td>
+                                            <td>
+                                                <audio controls>
+                                                <source src="sounds/{{ $word->sound_file }}" type="audio/mpeg">
```
Comment (nguyenthanhtung88) [inline resources/views/word/wordlist.blade.php]: Check indent

File: resources/views/word/wordlist.blade.php [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    <div class="container">
+        <div class="row">
+            <div class="col-md-10 col-md-offset-1">
+                <div class="panel panel-default">
+                    <div class="panel-heading">
+                        {{ trans('text.word') }}
+                    </div>
+                    <div class="panel-body">
+                    	{{ Form::open(['method' => 'POST', 'action' => 'WordController@search', 'files' => true, 'class' => 'form-inline']) }}
+                    		{{ Form::label('category_id', trans('text.category')) }}
```
Comment (nguyenthanhtung88) [inline resources/views/word/wordlist.blade.php]: Check indent

File: app/Http/Controllers/WordController.php [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+use App\Http\Requests\WordRequest;
+use App\Models\Category;
+use App\Models\Word;
+use App\Repositories\Contracts\WordRepositoryInterface as WordRepository;
+
+class WordController extends Controller
+{
+    public function __construct(WordRepository $wordRepository)
+    {
+        $this->wordRepository = $wordRepository;
```
Comment (jsnulla) [inline app/Http/Controllers/WordController.php]: Undefined variable `$wordRepository`

File: resources/views/word/create.blade.php [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+                    {{ Form::file('sound_file', ['class' => 'form-control']) }}
+                </div>
+
+                <div class="form-group">
+                    {{ Form::label('category_id', trans('text.category_name')) }}
+                    {{ Form::select('category_id', $categories, null, ['id' => 'category_id', 'class' => 'form-control']) }}
+                </div>
+
+                <div class="form-group">
+                    {{ Form::submit('Submit', ['class' => 'btn btn-success pull-right']) }}
```
Comment (jsnulla) [inline resources/views/word/create.blade.php]: use trans `Submit`

File: resources/views/word/edit.blade.php [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+                    {{ Form::file('sound_file', ['class' => 'form-control']) }}
+                </div>
+
+                <div class="form-group">
+                    {{ Form::label('category_id', trans('text.category_name')) }}
+                    {{ Form::select('category_id', $categories, $word->category->id, ['id' => 'category_id', 'class' => 'form-control']) }}
+                </div>
+
+                <div class="form-group">
+                    {{ Form::submit('Submit', ['class' => 'btn btn-success pull-right']) }}
```
Comment (jsnulla) [inline resources/views/word/edit.blade.php]: Use trans `Submit`

File: app/Http/Controllers/WordController.php [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $this->wordRepository->createOrUpdateWord($request);
+        return redirect('words');
+    }
+
+    public function edit(Word $word)
+    {
+        $categories = Category::lists('name', 'id');
+        $word_options = array(explode(',', $word->word_options));
+        $i = 1;
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/WordController.php]: What do you mean `$i` stand for ???

File: app/Http/Controllers/WordController.php [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    public function store(WordRequest $request)
+    {
+        $this->wordRepository->createOrUpdateWord($request);
+        return redirect('words');
+    }
+
+    public function edit(Word $word)
+    {
+        $categories = Category::lists('name', 'id');
+        $word_options = array(explode(',', $word->word_options));
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/WordController.php]: Variable name should be in `camelCase`

File: resources/views/word/create.blade.php [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+                    {{ Form::text('vietnamese_word', null, ['class' => 'form-control']) }}
+                </div>
+
+                <div class="form-group">
+                    {{ Form::label('japanese_word', trans('text.japanese_word')) }}
+                    {{ Form::text('japanese_word', null, ['class' => 'form-control']) }}
+                </div>
+
+                <div class="form-group">
+                    {{ Form::label('words_options', trans('text.words_option')) }}
```
Comment (nguyenthanhtung88) [inline resources/views/word/create.blade.php]: I think it should be
```
{{ Form::label('word_options', trans('text.word_options')) }}
```
it means options from specific word

File: resources/views/errors/form.blade.php [hunk 15]
```diff
@@ -1,6 +1,6 @@
- @if ($errors->any())
+@if ($errors->any())
     <ul class='alert alert-danger'>
-        @foreach ($errors->all() as $error)
+        @foreach($errors->all() as $error)
```
Comment (nguyenthanhtung88) [inline resources/views/errors/form.blade.php]: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#55-foreach
We need one space after `foreach` and before `(`

File: resources/views/word/create.blade.php [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+                </div>
+
+                <div class="form-group">
+                    {{ Form::label('japanese_word', trans('text.japanese_word')) }}
+                    {{ Form::text('japanese_word', null, ['class' => 'form-control']) }}
+                </div>
+
+                <div class="form-group">
+                    {{ Form::label('word_options', trans('text.word_options')) }}
+                    @for($i = 1; $i <= $countOfOption; $i++ )
```
Comment (nguyenthanhtung88) [inline resources/views/word/create.blade.php]: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#54-for

**Comment by nguyenthanhtung88** [comment]
OK

**Pull request merged**